classes/class.tapestry.php

```diff
@@ -267,6 +290,47 @@ public function isEmpty()
         return empty($this->rootId);
     }
 
+    public function validate($node, $parent = null)
+    {
+        $type = $node->tydeType;
+
+        if (!isset($type) || !is_string($type)) {
+            return true; // for backwards compatibility
+        }
+
+        if (!isset($parent)) {
+            return $type == "Module" || $type == "Regular";
+        }
+
+        $parentType = $parent->tydeType;
+        if ($parentType == "Module") {
+            return $type == "Stage";
+        } else if ($parentType == "Stage") {
+            return $type == "Question set";
+        } else if ($parentType == "Regular") {
+            return $type == "Module" || $type == "Regular";
+        } else {
+            // otherwise parent is a question set, so we shouldn't be able
```

Comment:
What if `$parentType` is not set? It should pass, no? (For backwards compatibility)